plugin/path_checkouts.go

```diff
@@ -5,11 +5,320 @@ import (
+			ttl = requestedTTL
+		case set.TTL > 0 && requestedTTL < set.TTL:
+			// The set's TTL isn't infinite and the caller requested a shorter TTL.
+			ttl = requestedTTL
+		}
+	}
+	due := time.Now().UTC().Add(ttl)
+	if ttl <= 0 {
+		// An unlimited lending period is allowed. For simplicity, set
+		// the due date to be 100 years in the future.
+		// The due date is mainly for use during "status" calls so it's
+		// easy to understand who has checked something out and for how long.
+		due = time.Now().UTC().Add(time.Hour * 24 * 365 * 100)
+	}
+	checkOut := &CheckOut{
+		IsAvailable:         false,
+		Due:                 due,
+		BorrowerEntityID:    req.EntityID,
+		BorrowerClientToken: req.ClientToken,
+	}
+
+	if err := b.checkOutHandler.CheckOut(ctx, req.Storage, availableServiceAccountName, checkOut); err != nil {
+		return nil, err
+	}
+	password, err := retrievePassword(ctx, req.Storage, availableServiceAccountName)
+	if err != nil {
+		return nil, err
+	}
+
+	respData := map[string]interface{}{
+		"service_account_name": availableServiceAccountName,
+		"password":             password,
+	}
+	internalData := map[string]interface{}{
+		"set_name":             setName,
+		"service_account_name": availableServiceAccountName,
+	}
+	resp := b.Backend.Secret(secretAccessKeyType).Response(respData, internalData)
+	resp.Secret.Renewable = true
+	resp.Secret.TTL = ttl
+	resp.Secret.MaxTTL = set.MaxTTL
+	return resp, nil
+}
+
+func (b *backend) secretAccessKeys() *framework.Secret {
+	return &framework.Secret{
+		Type: secretAccessKeyType,
+		Fields: map[string]*framework.FieldSchema{
+			"service_account_name": {
+				Type:        framework.TypeString,
+				Description: "Access Key",
+			},
+			"password": {
+				Type:        framework.TypeString,
+				Description: "Secret Key",
+			},
+		},
+		Renew:  b.renewCheckOut,
+		Revoke: b.endCheckOut,
+	}
+}
+
+func (b *backend) renewCheckOut(ctx context.Context, req *logical.Request, fieldData *framework.FieldData) (*logical.Response, error) {
+	setName := req.Secret.InternalData["set_name"].(string)
+	serviceAccountName := req.Secret.InternalData["service_account_name"].(string)
+
+	checkOut, err := b.checkOutHandler.Status(ctx, req.Storage, serviceAccountName)
+	if err != nil {
+		return nil, err
+	}
+	if checkOut.IsAvailable {
+		// It's possible that this renewal could be attempted after a check-in occurred either by this entity or by
+		// another user with access to the "manage check-ins" endpoint that forcibly checked it back in.
+		return logical.ErrorResponse(fmt.Sprintf("%s is already checked in, please call check-out to regain it", serviceAccountName)), nil
+	}
+
+	set, err := readSet(ctx, req.Storage, setName)
+	if err != nil {
+		return nil, err
+	}
+	if set == nil {
+		return logical.ErrorResponse(`"%s" doesn't exist`, setName), nil
+	}
+	// The new time due will be consistent with when the secret ttl ends.
+	checkOut.Due = time.Now().UTC().Add(req.Secret.TTL)
```

Comment:
Sometimes renewals are for less than the TTL. Like if a TTL is for 10 seconds, but there are only 2 seconds left before we hit the max TTL, then the lease TTL will be 2 seconds. But in this line of code, it would still think it's 10 seconds away. 

For that reason, perhaps we should strip out reflecting a time due, and simply let it all be handled by the leasing system. That would make it less cool to look up a service account's status - you'd only see whether it's available and the entity ID or client token it's checked out to. Not when it's due.

Open to thoughts here.

Reply:
Resolved by stripping "due".